## Oscar Data 수집

오스카 수상 예측 모델을 구현하기 위해 데이터를 수집, 정제했습니다.   
오스카를 대표하는 5개 부문 '작품상', '감독상', '각본상', '여우주연상', '남우주연상' 만 예측 할 예정입니다.   
kaggle에서 구한 `the_oscar_award.csv` 데이터를 이용했습니다.   


In [1]:
import pandas as pd
import numpy as np


In [2]:

odata = pd.read_csv("the_oscar_award.csv")
odata.head(20)

year_film  year_ceremony  ceremony                      category  \
0        1927           1928         1                         ACTOR   
1        1927           1928         1                         ACTOR   
2        1927           1928         1                       ACTRESS   
3        1927           1928         1                       ACTRESS   
4        1927           1928         1                       ACTRESS   
5        1927           1928         1                 ART DIRECTION   
6        1927           1928         1                 ART DIRECTION   
7        1927           1928         1                 ART DIRECTION   
8        1927           1928         1                CINEMATOGRAPHY   
9        1927           1928         1                CINEMATOGRAPHY   
10       1927           1928         1                CINEMATOGRAPHY   
11       1927           1928         1    DIRECTING (Comedy Picture)   
12       1927           1928         1    DIRECTING (Comedy Picture)   
13       1927           1928         1  DIRECTING (Dramatic Picture)   
14       1927           1928         1  DIRECTING (Dramatic Picture)   
15       1927           1928         1  DIRECTING (Dramatic Picture)   
16       1927           1928         1           ENGINEERING EFFECTS   
17       1927           1928         1           ENGINEERING EFFECTS   
18       1927           1928         1           ENGINEERING EFFECTS   
19       1927           1928         1           OUTSTANDING PICTURE   

                       name                 film    win  
0       Richard Barthelmess            The Noose  False  
1             Emil Jannings     The Last Command   True  
2            Louise Dresser      A Ship Comes In  False  
3              Janet Gaynor           7th Heaven   True  
4            Gloria Swanson       Sadie Thompson  False  
5             Rochus Gliese              Sunrise  False  
6   William Cameron Menzies             The Dove   True  
7              Harry Oliver           7th Heaven  False  
8             George Barnes     The Devil Dancer  False  
9            Charles Rosher              Sunrise   True  
10              Karl Struss              Sunrise   True  
11          Lewis Milestone  Two Arabian Knights   True  
12                Ted Wilde               Speedy  False  
13            Frank Borzage           7th Heaven   True  
14           Herbert Brenon      Sorrell and Son  False  
15               King Vidor            The Crowd  False  
16           Ralph Hammeras                       False  
17              Roy Pomeroy                Wings   True  
18         Nugent Slaughter                       False  
19        The Caddo Company           The Racket  False

 year_film, year_ceremony, ceremony, category, name, film, win 의 7개의 칼럼으로 이루어짐.

## category 리스트 확인 

In [3]:
odata["category"].unique()

array(['ACTOR', 'ACTRESS', 'ART DIRECTION', 'CINEMATOGRAPHY',
       'DIRECTING (Comedy Picture)', 'DIRECTING (Dramatic Picture)',
       'ENGINEERING EFFECTS', 'OUTSTANDING PICTURE',
       'UNIQUE AND ARTISTIC PICTURE', 'WRITING (Adaptation)',
       'WRITING (Original Story)', 'WRITING (Title Writing)', 'DIRECTING',
       'WRITING', 'OUTSTANDING PRODUCTION', 'SOUND RECORDING',
       'SHORT SUBJECT (Cartoon)', 'SHORT SUBJECT (Comedy)',
       'SHORT SUBJECT (Novelty)', 'ASSISTANT DIRECTOR', 'FILM EDITING',
       'MUSIC (Scoring)', 'MUSIC (Song)', 'DANCE DIRECTION',
       'WRITING (Screenplay)', 'ACTOR IN A SUPPORTING ROLE',
       'ACTRESS IN A SUPPORTING ROLE', 'SHORT SUBJECT (Color)',
       'SHORT SUBJECT (One-reel)', 'SHORT SUBJECT (Two-reel)',
       'IRVING G. THALBERG MEMORIAL AWARD', 'MUSIC (Original Score)',
       'CINEMATOGRAPHY (Black-and-White)', 'CINEMATOGRAPHY (Color)',
       'SPECIAL EFFECTS', 'ART DIRECTION (Black-and-White)',
       'ART DIRECTION (Color)', 'WR

category를 기준으로 5개 부문(작품상, 감독상, 각본상, 남우주연상, 여우주연상)만 선택  

작품상 = "BEST PICTURE" ->  year_ceremony 기준 1963 부터  
감독상 = "DIRECTING" -> year_ceremony 기준 1929 부터  
각본상 = "WRITING"   
 'WRITING (Story and Screenplay)', 'WRITING (Screenplay--Adapted)', 'WRITING (Screenplay--Original)',   
 'WRITING (Original Motion Picture Story)'  등등 각본상의 이름이 너무 많다 -> 'WRITING' 으로 통일시키는 작업 필요   
남우 주연상 = "ACTOR IN A LEADING ROLE" -> year_ceremony 기준 1977 부터   
여우 주연상 =  "ACTRESS IN A LEADING ROLE" -> year_ceremony 기준 1977 부터   

너무 오래전 영화는 다른 데이터를 구하기 어렵다.   
너무 최신 시상식으로만 예측한다면 데이터의 양이 부족하다.  

year_ceremony 기준으로 1977년 이후 데이터만 사용하기로 하자.  

## 각본상 category명을 'WRITING'으로 통일

In [4]:
odata["category"] = odata["category"].apply(lambda x: "WRITING" if x[0:7] == "WRITING" else x)

## 5개 부문 상만 선택

In [5]:
odata = odata[(odata["category"] == "ACTOR IN A LEADING ROLE") | 
              (odata["category"] == "ACTRESS IN A LEADING ROLE")| 
              (odata["category"] == "BEST PICTURE") | 
              (odata["category"] == "DIRECTING")|
              (odata["category"] == "WRITING")]

## year_ceremony 기준 1977년부터 데이터 

In [6]:
odata = odata[odata["year_ceremony"]>1976]
odata = odata.reset_index()
odata

index  year_film  year_ceremony  ceremony                 category  \
0      5291       1976           1977        49  ACTOR IN A LEADING ROLE   
1      5292       1976           1977        49  ACTOR IN A LEADING ROLE   
2      5293       1976           1977        49  ACTOR IN A LEADING ROLE   
3      5294       1976           1977        49  ACTOR IN A LEADING ROLE   
4      5295       1976           1977        49  ACTOR IN A LEADING ROLE   
...     ...        ...            ...       ...                      ...   
1324  10072       2018           2019        91                  WRITING   
1325  10073       2018           2019        91                  WRITING   
1326  10074       2018           2019        91                  WRITING   
1327  10075       2018           2019        91                  WRITING   
1328  10076       2018           2019        91                  WRITING   

                                                   name            film    win  
0                                        Robert De Niro     Taxi Driver  False  
1                                           Peter Finch         Network   True  
2                                    Giancarlo Giannini  Seven Beauties  False  
3                                        William Holden         Network  False  
4                                    Sylvester Stallone           Rocky  False  
...                                                 ...             ...    ...  
1324         Written by Deborah Davis and Tony McNamara   The Favourite  False  
1325                           Written by Paul Schrader  First Reformed  False  
1326  Written by Nick Vallelonga Brian Currie Peter ...      Green Book   True  
1327                          Written by Alfonso Cuarón            Roma  False  
1328                              Written by Adam McKay            Vice  False  

[1329 rows x 8 columns]

In [7]:
odata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1329 entries, 0 to 1328
Data columns (total 8 columns):
index            1329 non-null int64
year_film        1329 non-null int64
year_ceremony    1329 non-null int64
ceremony         1329 non-null int64
category         1329 non-null object
name             1329 non-null object
film             1329 non-null object
win              1329 non-null bool
dtypes: bool(1), int64(4), object(3)
memory usage: 74.1+ KB


In [8]:
odata.to_csv('odata_list.csv',index=False)

## Rottentomato API를 이용해 수집한 데이터


In [9]:
final= pd.read_csv("final.csv")
final

year_film  year_ceremony  ceremony                 category  \
0          1976           1977        49  ACTOR IN A LEADING ROLE   
1          1976           1977        49  ACTOR IN A LEADING ROLE   
2          1976           1977        49  ACTOR IN A LEADING ROLE   
3          1976           1977        49  ACTOR IN A LEADING ROLE   
4          1976           1977        49  ACTOR IN A LEADING ROLE   
...         ...            ...       ...                      ...   
1324       2018           2019        91                  WRITING   
1325       2018           2019        91                  WRITING   
1326       2018           2019        91                  WRITING   
1327       2018           2019        91                  WRITING   
1328       2018           2019        91                  WRITING   

                                                   name            film  \
0                                        Robert De Niro     Taxi Driver   
1                                           Peter Finch         Network   
2                                    Giancarlo Giannini  Seven Beauties   
3                                        William Holden         Network   
4                                    Sylvester Stallone           Rocky   
...                                                 ...             ...   
1324         Written by Deborah Davis and Tony McNamara   The Favourite   
1325                           Written by Paul Schrader  First Reformed   
1326  Written by Nick Vallelonga Brian Currie Peter ...      Green Book   
1327                          Written by Alfonso Cuar처n            Roma   
1328                              Written by Adam McKay            Vice   

        win         Gross Date  Runtime  ...  Musical  History  Action  Crime  \
0     False  2.826257e+07  Feb      114  ...      0.0      0.0     0.0    1.0   
1      True  2.368988e+07  Nov      121  ...      0.0      0.0     0.0    0.0   
2     False  1.240800e+04  Jan      116  ...      0.0      0.0     0.0    0.0   
3     False  2.368988e+07  Nov      121  ...      0.0      0.0     0.0    0.0   
4     False  1.172351e+08  Dec      120  ...      0.0      0.0     0.0    0.0   
...     ...           ...  ...      ...  ...      ...      ...     ...    ...   
1324  False  3.436678e+07  Dec      119  ...      0.0      1.0     0.0    0.0   
1325  False  3.448256e+06  May      113  ...      0.0      0.0     0.0    0.0   
1326   True  8.508017e+07  Nov      130  ...      0.0      0.0     0.0    0.0   
1327  False  5.986694e+07  Nov      135  ...      0.0      0.0     0.0    0.0   
1328  False  4.783628e+07  Dec      132  ...      0.0      0.0     0.0    0.0   

      Romance  Family  Adventure  Sport  Horror  War  
0         0.0     0.0        0.0    0.0     0.0  0.0  
1         0.0     0.0        0.0    0.0     0.0  0.0  
2         0.0     0.0        0.0    0.0     0.0  1.0  
3         0.0     0.0        0.0    0.0     0.0  0.0  
4         0.0     0.0        0.0    1.0     0.0  0.0  
...       ...     ...        ...    ...     ...  ...  
1324      0.0     0.0        0.0    0.0     0.0  0.0  
1325      0.0     0.0        0.0    0.0     0.0  0.0  
1326      0.0     0.0        0.0    0.0     0.0  0.0  
1327      0.0     0.0        0.0    0.0     0.0  0.0  
1328      0.0     0.0        0.0    0.0     0.0  0.0  

[1329 rows x 38 columns]

## "winnings" 칼럼 추가 
다른 시상식에서 수상 갯수를 의미하는 칼럼 추가 

###  중복된 영화 제목 제거

1329 ->645

In [10]:
final1 = final.drop_duplicates("film", keep="first")
final1.shape[0]

645

### selenium을 이용하여 구글 크롤링 

chromedriver.exe 를 설치해야 한다. 

In [11]:
# -*- coding: UTF-8 -*- 
# [ IMPORT ] #
import urllib.request
import requests
from bs4 import BeautifulSoup as bs
import pymysql
from  selenium import webdriver
from selenium.webdriver.common.keys import Keys
import time 

In [12]:
final1["winnings"] = None

C:\Users\kse02\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [13]:
driver = webdriver.Chrome("C://Users//kse02//Downloads//chromedriver_win32//chromedriver.exe")# Chromedriver PATH
driver.implicitly_wait(5)

for i in range(0,len(final1)):
    driver.get("https://www.google.co.kr/")
    search = final1.iat[i,final1.columns.get_loc('film')] + " awards" + " movie"
    driver.find_element_by_name('q').send_keys(search)
    driver.find_element_by_name('q').send_keys(Keys.RETURN)
   
    html = driver.page_source
    soup = bs(html, 'html.parser')
    if soup.find_all("div" , class_="title") == []:
        final1.iat[i, final1.columns.get_loc("winnings")]= None
    else :                                             
        final1.iat[i, final1.columns.get_loc("winnings")]= len(soup.find_all("div" , class_="title"))
    time.sleep(3)

WebDriverException: Message: chrome not reachable
  (Session info: chrome=80.0.3987.132)


In [ ]:
final1.to_csv('final1.csv',index=False)

In [14]:
final2= pd.read_csv("final1.csv")

### pandas.merge를 이용하여 winnings 칼럼 붙이기 

In [19]:
final3 = pd.merge(final,final2,how="left")
final3

year_film  year_ceremony  ceremony                 category  \
0          1976           1977        49  ACTOR IN A LEADING ROLE   
1          1976           1977        49  ACTOR IN A LEADING ROLE   
2          1976           1977        49  ACTOR IN A LEADING ROLE   
3          1976           1977        49  ACTOR IN A LEADING ROLE   
4          1976           1977        49  ACTOR IN A LEADING ROLE   
...         ...            ...       ...                      ...   
1324       2018           2019        91                  WRITING   
1325       2018           2019        91                  WRITING   
1326       2018           2019        91                  WRITING   
1327       2018           2019        91                  WRITING   
1328       2018           2019        91                  WRITING   

                                                   name            film  \
0                                        Robert De Niro     Taxi Driver   
1                                           Peter Finch         Network   
2                                    Giancarlo Giannini  Seven Beauties   
3                                        William Holden         Network   
4                                    Sylvester Stallone           Rocky   
...                                                 ...             ...   
1324         Written by Deborah Davis and Tony McNamara   The Favourite   
1325                           Written by Paul Schrader  First Reformed   
1326  Written by Nick Vallelonga Brian Currie Peter ...      Green Book   
1327                          Written by Alfonso Cuar처n            Roma   
1328                              Written by Adam McKay            Vice   

        win         Gross Date  Runtime  ...  Dec  Feb  Jan  Jul  Jun  Mar  \
0     False  2.826257e+07  Feb      114  ...    0    1    0    0    0    0   
1      True  2.368988e+07  Nov      121  ...    0    0    0    0    0    0   
2     False  1.240800e+04  Jan      116  ...    0    0    1    0    0    0   
3     False  2.368988e+07  Nov      121  ...    0    0    0    0    0    0   
4     False  1.172351e+08  Dec      120  ...    1    0    0    0    0    0   
...     ...           ...  ...      ...  ...  ...  ...  ...  ...  ...  ...   
1324  False  3.436678e+07  Dec      119  ...    1    0    0    0    0    0   
1325  False  3.448256e+06  May      113  ...    0    0    0    0    0    0   
1326   True  8.508017e+07  Nov      130  ...    0    0    0    0    0    0   
1327  False  5.986694e+07  Nov      135  ...    0    0    0    0    0    0   
1328  False  4.783628e+07  Dec      132  ...    1    0    0    0    0    0   

      May  Nov  Oct  Sep  
0       0    0    0    0  
1       0    1    0    0  
2       0    0    0    0  
3       0    1    0    0  
4       0    0    0    0  
...   ...  ...  ...  ...  
1324    0    0    0    0  
1325    1    0    0    0  
1326    0    1    0    0  
1327    0    1    0    0  
1328    0    0    0    0  

[1329 rows x 51 columns]

## Date 컬럼 더미변수로 변경

In [20]:
df = pd.get_dummies(final3['Date'])

In [21]:
final3 = pd.concat([final3,pd.get_dummies(final3['Date'])], axis=1)

## win 컬럼 T/F -> 1/0 으로 변경

In [22]:
final3['win'] = final3['win'].apply(lambda x : 1 if (x == True) else 0)

In [23]:
final3.head()

year_film  year_ceremony  ceremony                 category  \
0       1976           1977        49  ACTOR IN A LEADING ROLE   
1       1976           1977        49  ACTOR IN A LEADING ROLE   
2       1976           1977        49  ACTOR IN A LEADING ROLE   
3       1976           1977        49  ACTOR IN A LEADING ROLE   
4       1976           1977        49  ACTOR IN A LEADING ROLE   

                 name            film  win        Gross Date  Runtime  ...  \
0      Robert De Niro     Taxi Driver    0   28262574.0  Feb      114  ...   
1         Peter Finch         Network    1   23689877.0  Nov      121  ...   
2  Giancarlo Giannini  Seven Beauties    0      12408.0  Jan      116  ...   
3      William Holden         Network    0   23689877.0  Nov      121  ...   
4  Sylvester Stallone           Rocky    0  117235147.0  Dec      120  ...   

   Dec  Feb  Jan  Jul  Jun  Mar  May  Nov  Oct  Sep  
0    0    1    0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0    1    0    0  
2    0    0    1    0    0    0    0    0    0    0  
3    0    0    0    0    0    0    0    1    0    0  
4    1    0    0    0    0    0    0    0    0    0  

[5 rows x 63 columns]

In [ ]:
final3.to_csv('final1.csv',index=False)

## 5개의 부문으로 데이터 프레임 나눠 저장하기

In [24]:
final1= pd.read_csv("final1.csv")

In [26]:
final1_best = final1[final1['category']=='BEST PICTURE']
final1_best.reset_index(drop=True, inplace=True)
final1_best.to_csv('final1_best.csv',index=False)

final1_direct = final1[final1['category']=='DIRECTING']
final1_direct.reset_index(drop=True, inplace=True)
final1_direct.to_csv('final1_direct.csv',index=False)
                    
final1_write = final1[final1['category']=='WRITING']
final1_write.reset_index(drop=True, inplace=True)
final1_write.to_csv('final1_write.csv',index=False)
                    
final1_actor = final1[final1['category']=='ACTOR IN A LEADING ROLE']
final1_actor.reset_index(drop=True, inplace=True)
final1_actor.to_csv('final1_actor.csv',index=False)
                    
final1_actress = final1[final1['category']=='ACTRESS IN A LEADING ROLE']
final1_actress.reset_index(drop=True, inplace=True)
final1_actress.to_csv('final1_actress.csv',index=False)



In [36]:
final2['winnings']

0       11.0
1       14.0
2        NaN
3        NaN
4        7.0
        ... 
1324     NaN
1325     8.0
1326     NaN
1327     NaN
1328     NaN
Name: winnings, Length: 1329, dtype: float64

In [35]:
final['Awards']

0        21.0
1        16.0
2         2.0
3        16.0
4        17.0
        ...  
1324    177.0
1325     60.0
1326     53.0
1327    248.0
1328     32.0
Name: Awards, Length: 1329, dtype: float64

In [33]:
final1.isnull().sum()

year_film          0
year_ceremony      0
ceremony           0
category           0
name               0
film               0
win                0
Gross              0
Date               0
Runtime            0
Awards             0
nominations        0
imdb               0
tomato             0
Metacr             0
imdbVotes          0
Documentary        0
Thriller           0
Short              0
Comedy             0
Fantasy            0
Animation          0
Mystery            0
Music              0
Biography          0
Drama              0
Western            0
Sci-Fi             0
Musical            0
History            0
Action             0
Crime              0
Romance            0
Family             0
Adventure          0
Sport              0
Horror             0
War                0
winnings         801
Apr                0
Aug                0
Dec                0
Feb                0
Jan                0
Jul                0
Jun                0
Mar                0
May          